### DISCUSSION TITLE
Run time configuration JSON not working when triggering the DAG manually

### DISCUSSION THREAD

**Discussion by mytreyireddy** [apache/airflow#23718] [General]
Hello, Could anyone please help me with the below issue?

Created a model in DBT that has one SQL file such as
DBT-> Models->products.sql
` select * FROM products {% if is_incremental() %} WHERE partition_date >= 9 {% endif %}`

Configured DBT through airflow  and I want to pass the runtime parameters that would override the DBT commands in the airflow task.
But when I pass in arguments in the airflow UI : These commands are not being executed.
`{"dbt_command": ["dbt", "run", "--select", "products", "--full-refresh", "--target", "stage"]}`

```
def get_entry_cmd():
    incremental_cmds = [
        "dbt",
        "run",
        "--select",
        "products",
        "--target",
        "stage"]
    dag_run_conf = "{{ dag_run.conf}}"
    logger.info("Run time configuration parameters: ", dag_run_conf)
    if "dbt_command" in dag_run_conf:
        converted_dag_run=eval(dag_run_conf) # This converts string type to its equivalent dictionary
        cmds = converted_dag_run["dbt_command"]
        logger.info("Commands related to Full load/refresh or backfilling are configured:", cmds)
    else:
        cmds = incremental_cmds
        logger.info("Commands related to incremental load/refresh are configured:", cmds)
    return cmds

Dbt_Products_Task = KubernetesPodOperator(
    namespace="airflow",
    service_account_name="airflow",
    task_id="build-product-tables",
    dag = dag,
    cmds=get_entry_cmd(),
    is_delete_operator_pod=True,
    name=f"{DAG_NAME}.dag",
    secrets=[keys_env],
    configmaps=[f"{CHART_NAME}-configmap"],
    image=dag_vars["IMAGE"],
    in_cluster=True,
    get_logs=True,
    env_vars={
        "ENV": os.environ["ENV"],
    },
)

```
Could anyone give me inputs on this?  When I trigger the DAG , without passing any JSON configuration , the incremental_cmds are executed and its working. But when I pass the JSON configuration with full-refresh DBT commands, its not working.

**Comment by potiuk** [comment]
What do you mean by "not working"? Hard to read what's in your mind, but a description of what happens, what logs are generated, what runs and what does not would be helpful - try to help others to help you.

**Reply by mytreyireddy** [reply]
Hi Potiuk, 

Thank you for the response. I am mainly looking to  access  the run time JSON configuration parameters which is in jinja template form `"{{dag_run.conf.get('dbt_command', '') if dag_run }}"` inside the Kubernetes commands operator. Below is the airflow task.

```
Dbt_Products_Task = KubernetesPodOperator(
    namespace="airflow",
    service_account_name="airflow",
    task_id="build-product-tables",
    image="dbt-snowflake:latest",
    cmds = ["{{dag_run.conf.get('dbt_command', '') if dag_run}}"],
    provide_context=True,
    dag_run_conf_overrides_params=True,
    dag = dag,
    is_delete_operator_pod=True,
    name=f"{DAG_NAME}.dag",
    secrets=[keys_env],
    configmaps=[f"{CHART_NAME}-configmap"],
    in_cluster=True,
    get_logs=True,
    env_vars={
        "ENV": os.environ["ENV"],
    },
)
```

<img width="777" alt="Screen Shot 2022-05-19 at 1 05 35 AM" src="https://user-images.githubusercontent.com/92401998/169244313-ae6ae25d-ed9e-4f21-b562-0472331b8a5f.png">

The  json paramters which I gave on Airflow UI are not being accessible in cmds.  So if dag_run exists then the above parameters which are given in Airflow UI json box should be executed else the list of  pre-defined incremental commands  which is  incremental_cmds = [
        "dbt",
        "run",
        "--select",
        "products",
        "--target",
        "stage"] should be executed.

Could you please help me with this issue on how to pass the conditional statements `if else` inside the `cmds` in `kubernets operator` with `jinja` template ?

**Reply by potiuk** [reply]
``"{{ dag_run.conf}}"`` is only resolved if they are parameter of the operator's constructor. You cannot use it this way to to dynamically pass different values because your "get_cmds()" method is executed during parsing. you can only dynamically resolve jinja templates parameters if you pass them as strings. This will not work. You need to understand when your methods are called -  if you pass result of method call as constructor parameters, the parameters are not yet available.

What you can try to do is to extend KPO and write your own operator where you pass "{{dag_run.conf}} as parameter, and dynamically replace cmds in your execute method - but for that you need to look at the code of KPO and understand how it works under the hood.

Or you need to change your dag completely to conditionally run different tasks based on branch operator and have fixed set of commands in each task that you fill values of based on the "{{dag_run.conf}}"

Smth like:

```
task_run_when_conf_is_there = KubernetesPodOperator(
     ...
     cmds= ["dbt", "run", "--select", "{{dag_run.conf.select}}", ...) 

task_run_when_conf_is_missing = KubernetesPodOperator(

     cmds= ["dbt", "run","fixed", "parameters", "here") 

```

And add a branching operator based on dag_run.conf being set or not.

**Reply by mytreyireddy** [reply]
Thank you Potiuk! I really appreciate your help in taking time and giving me the approach to solve this runtime overrides problem. I will implement this approach and post the results in here.

**Reply by mytreyireddy** [reply]
Hi Potiuk,

I created a new DAG just to access overrides. Below are the commands that are passed in the overrides. 
`JSON on Airflow UI : {"dbt_command": ["products", "--full-refresh"]}`

<img width="841" alt="Screen Shot 2022-05-20 at 5 45 02 PM" src="https://user-images.githubusercontent.com/92401998/169627702-71a1a237-e25b-4d1c-8758-d3589edd7271.png">


The commands are now accessible but its in the format of list string. How could I remove the type string and list square braces? 
**Output in Airflow logs:** 
`
['dbt',
 'run',
 '--select',
 "['products', '--full-refresh']",
 '--target',
 'stage'] 
`
**Expected Output:**
`
['dbt',
 'run',
 '--select',
 'products', '--full-refresh',
 '--target',
 'stage'] 
`

**Original Code:**
```
Dbt_overrides_Reporting_Task = KubernetesPodOperator(
    namespace="airflow",
    service_account_name="airflow",
    task_id="overrides",
    #image=dag_vars["IMAGE"],
    image="dbt-snowflake:latest",
    cmds = ["dbt", "run", "--select", "{{dag_run.conf.get('dbt_command', '') if dag_run}}" , "--target", "stage"],
    provide_context=True,
    dag_run_conf_overrides_params=True,
    dag = dag,
    is_delete_operator_pod=True,
    name=f"{DAG_NAME}.dag",
    secrets=[keys_env],
    configmaps=[f"{CHART_NAME}-configmap"],
    in_cluster=True,
    get_logs=True,
    env_vars={
        "ENV": os.environ["ENV"],
    },
)
```

Could you please help me with this issue?

**Reply by potiuk** [reply]
In Airflow 2.2+ you have https://airflow.apache.org/docs/apache-airflow/2.2.4/concepts/operators.html#rendering-fields-as-native-python-objects which does exactly what you want.
